Concepts documentation for federated sidechains (#1247)
XRPLF/xrpl-dev-portal: content/concepts/consensus-network/federated-sidechains.md

---
html: federated-sidechains.html
parent: consensus-network.html
blurb: Learn about federated sidechains, which enable value in the form of XRP and other tokens (IOUs) to move efficiently between a sidechain and the XRP Ledger.
labels:
  - Blockchain
---
# Federated Sidechains

_Federated Sidechains are available as an Engineering Preview and can be used to develop and test using `rippled` 1.8.0._

> A sidechain is an independent ledger with its own consensus algorithm and transaction types and rules. It acts as its own blockchain. Federation enables value in the form of XRP and other tokens (IOUs) to move efficiently between a sidechain and the XRP Ledger mainchain (Devnet, Testnet, or Mainnet). Federated sidechains operate without compromising the impressive speed, efficiency, and throughput of the public Mainnet.

Reviewer: ```suggestion
A sidechain is an independent ledger with its own consensus algorithm and transaction types and rules. It acts as its own blockchain. Federation enables value in the form of XRP and other tokens to move efficiently between a sidechain and an XRP Ledger _mainchain_ (usually Mainnet, but could be [Testnet or Devnet](parallel-networks.html) for testing). Federated sidechains operate without compromising the speed, efficiency, and throughput of the public Mainnet.
```

- No need to use the term "IOU"
- Make it clearer that this is the definition of "mainchain"
- Avoid subjective terms like "impressive" that come across as marketing hype

> Federated sidechains enable the XRP Ledger community developers to launch new features and innovative applications built on foundational XRP Ledger technology. You can customize the core XRP Ledger technology to the needs of a specific use case or project and run it as its own blockchain. Here are a few examples:

Reviewer: ```suggestion
Federated sidechains enable developers to launch new features and innovative applications using the foundation of XRP Ledger technology. Sidechains can customize the XRP Ledger protocol to the needs of a specific use case or project and run it as its own blockchain. Here are a few examples:
```

> * Build a smart contract layer like EVM smart contract, a web assembly smart contract, or a Move VM smart contract. 

Reviewer: ```suggestion
* Build a smart contract layer, powered by an engine compatible with the Ethereum Virtual Machine (EVM), web assembly, or a Move VM. 
```

* Build your own algorithmic stable coin with customised ledger types and transaction rules.

> * Build permissioned or nearly permissionless, centralized or largely decentralized ledgers whose assets can be traded on the Mainnet DEX

Reviewer: ```suggestion
* Build permissioned or nearly permissionless, centralized or largely decentralized ledgers whose assets can be traded on the Mainnet [decentralized exchange](decentralized-exchange.html).
```

- Add a period for consistency
- Avoid the abbreviation DEX which isn't defined elsewhere

> * Create private or public sidechains, and possibly make the public sidechains available to the community to leverage various use cases. For example, a [smart sidechain with Hooks](https://hooks-testnet.xrpl-labs.com/) enabled.

Reviewer: I think this is redundant with the above bullet points as written. (Hooks are essentially smart contracts, and the previous point about permissioned or permissionless networks applies to private or public as well.)

Author: Fixed.

> ## How Do Federated Sidechains Work

Reviewer: ```suggestion
## How Federated Sidechains Work
```

> A sidechain is an independent ledger with its own consensus algorithm and transaction types and rules. Each sidechain runs its own set of servers (nodes and validators) and does not rely on the validators on the Mainnet to submit transactions on the sidechain.

Reviewer: ```suggestion
A sidechain is an independent ledger with its own consensus algorithm and transaction types and rules. Each sidechain is run by its own set of servers (including validators) and does not rely on the validators on the Mainnet for transactions on the sidechain.
```

- The network doesn't run the servers; it's the other way around.
- Avoid the term "node," which is overloaded to mean too many different hings.
- Validators don't submit transactions. They only put them in order (and confirm that they got the same results).

> Each sidechain has two door accounts, one each on mainchain and sidechain, that are controlled by the federators on the sidechain.The federators independently listen for transactions on these door accounts.

Reviewer: ```suggestion
Each sidechain has two door accounts, one on the sidechain and one on the mainchain, that are controlled by the federators on the sidechain. The federators listen for transactions to and from both of these door accounts.
```

- Add missing space
- At first I misread "one each on mainchain and sidechain" as "one on each...". I think it's confusing because of the way the sentence is introduced with "Each sidechain..." Rephrased to make that part read better.
- I don't think the word "independently" is needed here.
- Transactions "to and from" is probably more intuitive than "on" and means the same thing in this context.

> Federators, similar to validators on the mainchain but only live on the sidechain, jointly control the door accounts using the built-in multi-signature scheme available on the XRP Ledger. The door account has a signers list which consists of signing keys of all federators on the sidechain. 

Reviewer: To my understanding, the following is more correct and also more concise:

```suggestion
The sidechain has _federators_ who jointly control the door accounts on both networks using [multi-signing](multi-signing.html) so that 80% of federators must approve a transaction. In many cases, the federators should also be the trusted validators of the sidechain.
```

> When a door account receives a transaction, the federator signs the transaction and broadcasts this information to other federators on the sidechain.

Reviewer: This is missing or mixing up some important details.

```suggestion
When a door account receives a transaction on either the sidechain or the mainchain, the federators create a mirror transaction on the other chain. (For example, if you send XRP _to_ the mainchain door account, the federators create a transaction on the sidechain to send XRP _from_ the sidechain door account to the intended recipient.) The federators sign the transaction and broadcast it to each other.
```

> Simultaneously, federators also listen for signed transactions from other federators and collect them.

Reviewer: If you want a paragraph break, be sure to leave a blank line in between. Otherwise, a single line break like this gets treated as hard-wrapping. (Don't hard-wrap lines for XRPL docs, either.)

> A transaction is valid only when 80% of the signers on the list have signed. A quorum is reached when 80% of federators have signed and the transaction is submitted to the door account and an asset is issued to the destination.

Reviewer: - Not all transactions issue an asset.
- Lots of passive voice here.
- The beginning of the second sentence is redundant with the first sentence.

```suggestion
When 80% of the federators have signed the transaction, they submit it to the sidechain or mainchain as appropriate. This way, assets that the mainchain door account holds can be allocated to others on the sidechain, and assets that sidechain door account receives can be sent to others on the mainchain.
```

> When an asset is issued from the mainchain to the sidechain, that asset is locked up in the door account. The asset is only unlocked on the mainchain when a transaction from the sidechain to the main chain is submitted. Transactions within the sidechain are not visible to the servers on the sidechain. 

Reviewer: With my proposed edits above, I think you can remove the first two sentences (which have more passive voice anyway). 

As for the last sentence, I think there's a typo with "sidechain" twice; as written, it doesn't make sense. I believe the intended meaning is that the mainchain does not see the sidechain transactions; within the sidechain, assets can be held by different parties and change hands frequently, but as far as the mainchain is concerned, the door account was holding those same assets the whole time.

## Terminology

> _Sidechain_: A sidechain is an independent ledger which provides a way to transfer assets from the XRP Ledger to the sidechain, and return assets from the sidechain back to the XRP Ledger. Each sidechain has its own consensus algorithm and transaction types and rules. 

Reviewer: ```suggestion
_Sidechain_: An XRP Ledger sidechain is another blockchain based on XRP Ledger technology. A _federated_ sidechain provides a way to transfer assets from a mainchain to the sidechain and back. A sidechain can use an exact copy of the XRP Ledger's protocol or it can make changes, including to the [consensus algorithm](consensus.html), [transaction types](transaction-types.html), and other rules.
```

- Rephrased to be clearer what it means by having its own consensus algorithm
- Not sure if this is a useful distinction, but I rephrased to be a more precise definition of what makes a "sidechain" and what makes one "federated".
- Added some relevant links.

 
_Federator_: A server on the sidechain that listens for triggering transactions on both the main chain and the side chain. Each federator has a signing key associated with it that is used to sign transactions. A transaction must be signed by a quorum of federators before it can be submitted. Federators are responsible for creating and signing valid response transactions, collecting signatures from other federators, and submitting transactions to the main chain and side chain.

> _Main chain_: XRP Ledger where assets originate and where assets will be locked while being used on the side chain. For most applications, the main chain will be the XRP Ledger Devnet, Testnet, or Mainnet.

Reviewer: ```suggestion
_Mainchain_: The blockchain where assets originate and where assets are held while being used on the side chain. For most sidechains, the mainchain is the XRP Ledger Mainnet, Testnet, or Devnet.
```

- Use "mainchain" (one word) for consistency with the rest of the article
- Avoid future tense.

> _Side chain_: Custom independent ledger where proxy assets for the locked main chain assets are issued. Side chains may have rules, transactors, and validators that are very different from the main chain. Proxy assets on the side chain can be sent back to the main chain where they will be unlocked from the control of the federators.

Reviewer: ```suggestion
_Sidechain_: A custom blockchain where some assets are backed by a mainchain. Proxy assets are issued in the sidechain, with the equivalent assets held by a door account on the mainchain. Sidechains have separate history, rules, and validators than the mainchain. Proxy assets on the sidechain can be sent back to the mainchain and unlocked from the control of the federators.
```

> _Door account_: An account controlled by the federators. There are two door accounts: one on the main chain and one on the side chain. Cross chain transactions are started by users sending assets to a door account. Main chain to side chain transactions cause the balance to increase on the main chain door account and the balance to decrease on the side chain door account. It is called a "door" because it is the mechanism to move assets from one chain to another - much like going between rooms in a house requires stepping through a door.

Reviewer: ```suggestion
_Door account_: An account controlled by the federators. There are two door accounts: one on the mainchain and one on the sidechain. Cross chain transactions are started by users sending assets to a door account. Mainchain to sidechain transactions cause the balance to increase on the mainchain door account and the balance to decrease on the sidechain door account. It is called a "door" because it is the mechanism to move assets from one chain to another—much like going between rooms in a house requires stepping through a door.
```

- "mainchain" and "sidechain" as one word, for consistency
- use a real emdash

> _Triggering transaction_: A transaction that causes the federators to start the process of signing and submitting a new response transaction. For example, sending XRP to the main chain's door account is a triggering transaction that will cause the federators to submit a new transaction on the side chain.

Reviewer: ```suggestion
_Triggering transaction_: A transaction that causes the federators to start the process of signing and submitting a new response transaction. For example, sending XRP to the mainchain's door account is a triggering transaction that causes the federators to submit a new transaction on the sidechain.
```

> _Response transaction_: A transaction submitted by the federators in reaction to a triggering transaction. Note that triggering transaction and response transaction depends on context. Sending XRP from a door account to a user account is a response transaction when thinking about cross chain transactions. It is a triggering transaction when thinking about how to handle failed transactions.

Reviewer: ```suggestion
_Response transaction_: A transaction submitted by the federators in reaction to a triggering transaction. In most cases, the response transaction occurs on the opposite chain as the triggering transaction. However, there are some exceptions for handling failed transactions.
```

## How to Set Up a Federated Sidechain

> Federated Sidechains are currently available as an Engineering Preview so you can experiment and explore the potential of sidechains. You can connect sidechains to the XRP Ledger Testnet, Devnet, or Mainnet as soon as network servers upgrade to XRPL 1.8.

Reviewer: ```suggestion
Federated Sidechains are currently available as an Engineering Preview so you can experiment and explore the potential of sidechains. You can connect sidechains to the XRP Ledger Testnet, Devnet, or Mainnet as long as [the servers](the-rippled-server.html) in the mainchain network are running version 1.8.0 or higher.
```

- Reworded to stay relevant after the 1.8.0 release is old news
- Added relevant link for the server

**Caution:** You can connect sidechains to the XRP Ledger Mainnet to develop and test with small amounts; it is not recommended for production use cases until federated sidechains are available in a release. 

Setting up a sidechain involves the following high-level steps:

> 1. Clone the rippled sidechain branch: https://github.com/ripple/rippled/tree/sidechain.

Reviewer: ```suggestion
1. Clone the `rippled` source code and check out the `sidechain` branch: https://github.com/ripple/rippled/tree/sidechain.
```

> 2. Write custom transactors for your sidechain. Note that this is an important and non-trivial task. Refer to the XRP Ledger Protocol Reference for information about transaction types.

Reviewer: ```suggestion
2. Customize the source code for your sidechain. For example, you may want to write custom [transaction types](transaction-types.html). Note that this is an important and non-trivial task.
```

- Slightly more generic and optional. Custom transactors are not a strict requirement for a sidechain after all.
- Link transaction types instead of using text to send people to the reference.

> For example, the XRPL Labs Hooks project uses custom transactors. 

Reviewer: ```suggestion
```

This detail is not necessary. Also, putting a line like this in between numbered points may break the list formatting.

3. Each sidechain federator has its own configuration file that must be updated to include the following information:  

>     - [sidechain] stanza - add details such as signing key, main chain account, and the main chain address (IP and port) to listen to. 

Reviewer: ```suggestion
    - `[sidechain]` stanza - add details such as signing key, mainchain account, and the mainchain address (IP and port) to listen to. 
```

>     - [sidechain_assets] stanza - define assets that can be used for cross-chain transactions (XRP or IOU), exchange rate for the assets, and optional refund penalty to discourage transactions that may default.

Reviewer: ```suggestion
    - `[sidechain_assets]` stanza - define assets that can be used for cross-chain transactions (XRP or [issued tokens](issued-currencies.html)), exchange rate for the assets, and optional refund penalty to discourage transactions that may default.
```

    - [sidechain_federators] stanza - list of federators public keys to be used for signing. This list is common for all federators on the sidechain.

> 4. Setup “Door” accounts to enable cross chain transactions which involves the following tasks:

Reviewer: ```suggestion
4. Set up door accounts to enable cross chain transactions. This involves the following steps (on _both_ chains):
```

    - Create and fund the door accounts. 

>     - Set the signers list for the door accounts.

Reviewer: - Should be "SignerList", not "signers list"
- Link to [Set up multi-signing](set-up-multi-signing.html)

>     - Reserve 3 tickets for error handling.

Reviewer: Link [Tickets](https://xrpl.org/tickets.html) and/or [Use Tickets](https://xrpl.org/use-tickets.html).

Author: Linked to Tickets.

>     - And finally, disable the master key to the door account to ensure that the federators jointly control the door account. 

Reviewer: - Link [Disable Master Key Pair](https://xrpl.org/disable-master-key-pair.html)
- Say "key pair" instead of "key" to be more accurate.

Author: Updated.

>     Note that it is important to perform this final step only after successful completion of the previous steps.

Reviewer: I believe this is supposed to be a continuation of the previous bullet point about disabling the master key. However, it isn't formatted that way. You need to include a blank line above it, and indent this line an additional 4 spaces to make it formatted as a child of the nested bullet points.

Author: Done.

> The _Sidechain Launch Kit_ is a CLI tool that simplifies setting up federated sidechains and can be used to quickly spin up a sidechain on your local machine. The launch kit also installs the interactive sidechain_shell (RiplRepl) that enables you to interact with these sidechains.
> 
> Follow the instructions on https://github.com/xpring-eng/sidechain-launch-kit/blob/main/README.md to install and play with sidechains.

Reviewer: ```suggestion
The _Sidechain Launch Kit_ is a commandline tool that simplifies setting up federated sidechains and can be used to quickly spin up a sidechain on your local machine. The launch kit also installs an interactive sidechain_shell (RiplRepl) that enables you to interact with the sidechain.

[Sidechain Launch Kit >](https://github.com/xpring-eng/sidechain-launch-kit/blob/main/README.md)
```

This should make the link appear as a button on xrpl.org so it really comes across as a call to action.

## See Also

- **Concepts:**

>     - [Federated Sidechains Video](https://www.youtube.com/embed/UpVjO68tLIo)

Reviewer: I hear this is the new/preferred link: https://www.youtube.com/embed/NhH4LM8NxgY

(For the future, I think we're working on a way to more elegantly embed these types of videos earlier / more prominently without being annoying.)

Author: Updated.

    - [sidechain_federators] stanza - list of federators public keys to be used for signing. This list is common for all federators on the sidechain.
4. Set up door accounts to enable cross chain transactions. This involves the following steps (on _both_ chains):
    - Create and fund the door accounts. 

>     - Set the ["SignerList"](set-up-multi-signing.html) for door accounts.

Reviewer: ```suggestion
    - [Set up the SignerList](set-up-multi-signing.html) for the door account.
```

- We don't use quotation marks for ledger object types such as SignerList.
- These bullet points are introduced as being needed on both chains. Since there's only one door account per chain, it should be singular here.
- The link more directly matches where it takes you to (["Set up Multi-signing" tutorial](https://xrpl.org/set-up-multi-signing.html)) if you include "Set up the" in the link text. If the link text is only "SignerList" then I would expect it to go to the [SignerList object reference](https://xrpl.org/signerlist.html).

>     - Reserve three(3) [tickets](tickets.html) for error handling.

Reviewer: ```suggestion
    - Create three [tickets](tickets.html) for error handling.
```

I suggest "Create" instead of "Reserve" since the transaction type is TicketCreate and the "Reserve" is something different.

As for the numeral or spelled-out number thing, the Microsoft style guide has this to say:

> In body text, use numerals for 10 and greater. Spell out zero through nine if the number does not precede a unit of measure or is not used as input. [Good examples: 3 centimeters, five databases; bad examples: 2 disks, twelve functions].

You could argue that "3" here is going to be used as input to a TicketCreate transaction but it's indirect enough here that I think spelling it out is appropriate. I would be OK with either one, but doing both with the parentheses here is almost certainly more confusing.
